src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -16,8 +18,12 @@ namespace Microsoft.AspNet.Mvc
     /// </summary>
     public class SessionStateTempDataProvider : ITempDataProvider
     {
-        private static JsonSerializer jsonSerializer = new JsonSerializer();
-        private static string TempDataSessionStateKey = "__ControllerTempData";
+        private const string TempDataSessionStateKey = "__ControllerTempData";
+        private readonly JsonSerializer _jsonSerializer = JsonSerializer.Create(
+            new JsonSerializerSettings()
+            {
+                TypeNameHandling = TypeNameHandling.Auto
```

Comment:
Need to consult with @blowdart to use this because it is generally not allowed to use any `TypeNameHandling` value other than `None`. I don't know if we can guarantee that the data is coming from a trusted source. For example, suppose that the session state provider is a custom one that stores all session data in a cookie - that is no longer trusted data because it can be tampered. Our default session state providers can't be tampered because they are stored in trusted environments (local machine memory, Azure Cache, etc.).